File: frontend/src/categoryGroups.js
// categoryGroups.js
//
// Real, presentation-only grouping of the 18 fine-grained categories
// (backend/core/categorize.py's own CATEGORIES, unchanged — this file
// groups them, it never reclassifies an agent) into 5 top-level groups, so
// the category filter isn't 18-19 flat, same-weight chips. Grouping is
// based on what actually exists in the real category set today, not a
// generic taxonomy forced onto it — e.g. there's no group for a category
// this project doesn't have.
//
// 'Unclassified' is deliberately left OUT of every group, kept as its own
// separate bucket — it already has its own dedicated "Hide/Show
// unclassified" toggle elsewhere in the UI, and folding it into a group
// (e.g. "Content & Community") would misleadingly imply a real
// classification that was never made.
//
// Shared by web + mobile so the grouping can't silently drift between them.
export const CATEGORY_GROUPS = [
  {
    id: 'trading-defi',
    label: 'Trading & DeFi',
    categories: ['Grid Trading', 'Rebalancing', 'Yield Optimisation', 'Health Factor Monitoring', 'Trading Signals', 'Copy Trading'],
  },
  {
    id: 'data-analysis',
    label: 'Data & Analysis',
    categories: ['Data Analysis', 'Research', 'Prediction Markets'],
  },
  {
    id: 'security-trust',
    label: 'Security & Trust',
    categories: ['Smart Contract Auditing', 'Identity & Verification'],
  },
  {
    id: 'content-community',
    label: 'Content & Community',
    categories: ['Content & Copywriting', 'Social & Community', 'Customer Support', 'NFT & Generative Art', 'Gaming'],
  },
  {
    id: 'payments-infra',
    label: 'Payments & Infrastructure',
    categories: ['Payments & Settlement', 'Developer Tools'],
  },
];

// Real reverse lookup: fine category name -> its group id (or null for
// 'Unclassified' / anything not in any group above).
export const CATEGORY_TO_GROUP = CATEGORY_GROUPS.reduce((map, group) => {
  for (const cat of group.categories) map[cat] = group.id;
  return map;
}, {});

export function groupForCategory(category) {
  return CATEGORY_TO_GROUP[category] || null;
}

export function groupLabel(groupId) {
  return CATEGORY_GROUPS.find((g) => g.id === groupId)?.label || groupId;
}

File: backend/core/category_groups.py
"""
category_groups.py

The real, backend-side mirror of frontend/src/categoryGroups.js's own
CATEGORY_GROUPS — same real 5-group presentation grouping of
categorize.py's 18 fine-grained categories, kept in sync manually (no
cross-language shared module for this one small, stable mapping — same
real tradeoff already made for core/canary.py's own real ALLOWED_GROUPS
scope, which used to keep a private, second, partial copy of this exact
mapping before this module existed).

Real fix (2026-08-28): this used to be duplicated — core/canary.py had
its own private `_GROUP_CATEGORIES` covering only 2 of the 5 real groups
(canary's own real scope never needed the rest). core/pnl.py needs the
"Trading & DeFi" group specifically too, which would have made a third,
independent copy. Pulled out into this one real, shared module instead;
canary.py now imports from here rather than keeping its own.

'Unclassified' is deliberately not a key here, same real reasoning as the
frontend's own file: it's not a real classification, so it isn't a group.
"""

CATEGORY_GROUPS: dict[str, set[str]] = {
    "trading-defi": {"Grid Trading", "Rebalancing", "Yield Optimisation", "Health Factor Monitoring", "Trading Signals", "Copy Trading"},
    "data-analysis": {"Data Analysis", "Research", "Prediction Markets"},
    "security-trust": {"Smart Contract Auditing", "Identity & Verification"},
    "content-community": {"Content & Copywriting", "Social & Community", "Customer Support", "NFT & Generative Art", "Gaming"},
    "payments-infra": {"Payments & Settlement", "Developer Tools"},
}


def group_for_category(category: str | None) -> str | None:
    """Real reverse lookup — the fine category's real group id, or None
    for 'Unclassified' / anything not in any group above."""
    if not category:
        return None
    for group_id, categories in CATEGORY_GROUPS.items():
        if category in categories:
            return group_id
    return None
